### PR TITLE
Update //build for cobalt for updated //base.

### PR DESCRIPTION
**Pull request by ghost**
b/323994356

### CODE REVIEW

File: base/BUILD.gn [hunk 1]
```diff
@@ -3535,6 +3559,19 @@ test("base_unittests") {
       "process/process_metrics_unittest.cc",
       "process/process_unittest.cc",
       "process/process_util_unittest.cc",
+
+      "allocator/dispatcher/dispatcher_unittest.cc",
```
Comment (sherryzy) [inline base/BUILD.gn:3563]: Can we add a comment here for why these tests need to be disabled?

File: base/containers/checked_iterators.h [hunk 2]
```diff
@@ -267,6 +269,27 @@ struct pointer_traits<::base::CheckedContiguousIterator<T>> {
 };
 
 }  // namespace std
+#elif defined(STARBOARD)
```
Comment (sherryzy) [inline base/containers/checked_iterators.h:272]: I'm not sure this is necessary. See the upstream fix: https://gcc.gnu.org/pipermail/libstdc++-cvs/2020q2/031260.html. But it's ok to use it as a workaround for now. 
Can you create a TODO for removing this when we upgrade to c++20?

File: base/strings/string_util_starboard.h [hunk 3]
```diff
@@ -46,14 +46,6 @@ inline int vsnprintf(char* buffer, size_t size,
   return SbStringFormat(buffer, size, format, arguments);
 }
 
-inline int strncmp16(const char16* s1, const char16* s2, size_t count) {
```
Comment (sherryzy) [inline base/strings/string_util_starboard.h:49]: Why do we remove this?

File: base/values.cc [hunk 4]
```diff
@@ -877,6 +877,36 @@ size_t Value::List::size() const {
   return storage_.size();
 }
 
+#if defined(STARBOARD)
```
Comment (sherryzy) [inline base/values.cc:880]: The same as above. Can we add a TODO to remove this.

File: build/config/compiler/BUILD.gn [hunk 5]
```diff
@@ -1774,14 +1804,38 @@ config("chromium_code") {
...
+    if (!is_starboard) {
+      cflags += [
+        # Warn on unnecessary extra semicolons outside of function definitions.
+        "-Wextra-semi",
+      ]
+    }
+
+    # Suppress warning in old //net.
+    if (is_starboard) {
+      cflags += [
```
Comment (sherryzy) [inline build/config/compiler/BUILD.gn:1818]: Why do we need this change? Are the warnings caused by this build/config change?

File: build/toolchain/win/msvc_toolchain.gni
Comment (sherryzy) [inline build/toolchain/win/msvc_toolchain.gni:1]: For all these toolchain changes, we would better have @andrewsavage1 to review it as well..

**Review by sherryzy — COMMENTED** [review]
This PR is doing multiple things together, eg. fixing some unit tests, updating build config. For the build config part change, I don't have much context and would love @andrewsavage1 to review it as well. Before he come back, can we split this PR to a two smaller ones, one for fixing base unit tests and one for updating build configs. We can first submit the unit tests fixes and make the builds green, then update build config, finally clean up those unnecessary changes like some Macro for if defined(STARBOARD). In that way, it's much more clearer what unit tests fails due to updating build config.

**Comment by andrewsavage1** [comment]
> This PR is doing multiple things together, eg. fixing some unit tests, updating build config.

I agree with Sherry—is there a reason it's necessary for `//base` to be changed in this PR or can `//build` be updated to what it is in trunk alone? To me it looks like the `//base` changes aren't related

If `//build` is being updated to what's in trunk that seems fine, though with #2382 merged we'll want to ensure we maintain all targets building on linux

**Review by sherryzy — APPROVED** [review]

**Comment by andrewsavage1** [comment]
> `gn check` returns no errors and default target compiles on linux devel.

It looks like linux/devel is still failing to build here?

File: base/allocator/dispatcher/internal/dispatch_data.h [hunk 6]
```diff
@@ -6,7 +6,9 @@
 #define BASE_ALLOCATOR_DISPATCHER_INTERNAL_DISPATCH_DATA_H_
 
 #include "base/allocator/buildflags.h"
+#if defined(USE_HACKY_COBALT_CHANGES)
```
Comment (andrewsavage1) [inline base/allocator/dispatcher/internal/dispatch_data.h:9]: Should this be a !defined as it's in upstream code already?

File: base/allocator/partition_allocator/BUILD.gn [hunk 7]
```diff
@@ -360,6 +360,11 @@ component("partition_alloc") {
   if (enable_pkeys && is_debug) {
     configs += [ ":no_stack_protector" ]
   }
+
+  # Need to include |component("partition_alloc")| for gn check.
```
Comment (andrewsavage1) [inline base/allocator/partition_allocator/BUILD.gn:364]: Prefer using one of ` check_includes = false`/`friend`/`nogncheck`

File: base/allocator/partition_allocator/partition_alloc_base/threading/platform_thread.h [hunk 8]
```diff
@@ -18,7 +18,9 @@
 #include "base/allocator/partition_allocator/partition_alloc_base/time/time.h"
 #include "build/build_config.h"
 
-#if BUILDFLAG(IS_WIN)
+#if defined(STARBOARD)
```
Comment (andrewsavage1) [inline base/allocator/partition_allocator/partition_alloc_base/threading/platform_thread.h:21]: This is still getting included even though we're not building with partition_alloc?

File: base/files/file_util.cc [hunk 9]
```diff
@@ -219,6 +219,24 @@ bool ContentsEqual(const FilePath& filename1, const FilePath& filename2) {
   // We open the file in binary format even if they are text files because
   // we are just comparing that bytes are exactly same in both files and not
   // doing anything smart with text formatting.
+#ifdef STARBOARD
+  // std::ifstream doesn't work on all our platforms.
```
Comment (andrewsavage1) [inline base/files/file_util.cc:223]: Do we know that this is still true with our newer c++ standard requirements? If it's something we should come back to check, change the macro check to USE_HACKY_COBALT_CHANGES (probably with a comment to clarify what needs to be checked)

File: base/files/file_util.cc [hunk 10]
```diff
@@ -252,8 +270,10 @@ bool ContentsEqual(const FilePath& filename1, const FilePath& filename2) {
   file1.close();
   file2.close();
   return true;
+#endif
 }
 
+#if !defined(STARBOARD)
```
Comment (andrewsavage1) [inline base/files/file_util.cc:276]: Same as above & with other instances in this file and the header

File: base/metrics/persistent_histogram_storage.cc [hunk 11]
```diff
@@ -30,7 +30,8 @@ constexpr size_t kAllocSize = 1 << 20;  // 1 MiB
 void* AllocateLocalMemory(size_t size) {
   void* address;
 
-#if BUILDFLAG(IS_WIN)
+#if defined(STARBOARD)
```
Comment (andrewsavage1) [inline base/metrics/persistent_histogram_storage.cc:33]: Long term is it okay to have a noop here? If we should clean it up, switch it to USE_HACKY_COBALT_CHANGES

Same thing with base/metrics/persistent_memory_allocator.cc

File: base/test/test_suite.cc [hunk 12]
```diff
@@ -4,7 +4,9 @@
 
 #include "base/test/test_suite.h"
 
+#ifndef STARBOARD
 #include <signal.h>
```
Comment (andrewsavage1) [inline base/test/test_suite.cc:8]: Why can't we include this with starboard?

File: build/config/compiler/BUILD.gn (hunk 5 above)
Comment (andrewsavage1) [reply build/config/compiler/BUILD.gn:1818]: Considering that this seems to be specifically for the old //net, let's add a TODO here to come back and try to remove them later

File: net/base/filename_util_internal.cc [hunk 13]
```diff
@@ -209,7 +209,7 @@ bool FilePathToString16(const base::FilePath& path, std::u16string* converted) {
 #if BUILDFLAG(IS_WIN)
   converted->assign(path.value().begin(), path.value().end());
   return true;
-#elif BUILDFLAG(IS_POSIX) || BUILDFLAG(IS_FUCHSIA)
+#elif BUILDFLAG(IS_POSIX) || BUILDFLAG(IS_FUCHSIA) || defined(USE_HACKY_COBALT_CHANGES)
```
Comment (andrewsavage1) [inline net/base/filename_util_internal.cc:212]: Is this not the code we want to run long term? What's wrong with it?

File: base/allocator/partition_allocator/BUILD.gn (hunk 7 above)
Comment (andrewsavage1) [reply base/allocator/partition_allocator/BUILD.gn:364]: Could you add a bug to the taskflow backlog to clean this up and add that to the TODO please?

**Review by andrewsavage1 — APPROVED** [review]

**Pull request merged**